### ISSUE TITLE
Sphinx crashes with pickling error when run with multiple processes and remote image

### ISSUE THREAD

**Issue by simeonf** [sphinx-doc/sphinx#4279]
Subject: When building with multiple processes image directives can cause an infinite loop

### Problem

Building a trivial sphinx docs site results in crash in the worker and the parent process never returns and must be interrupted.

#### Procedure to reproduce the problem
```
$ sphinx-build -j 2 -b html source build
Running Sphinx v1.6.5
making output directory...
loading pickled environment... not yet created
building [mo]: targets for 0 po files that are out of date
building [html]: targets for 2 source files that are out of date
updating environment: 2 added, 0 changed, 0 removed
reading sources... [ 50%] foo
reading sources... [100%] index

looking for now-outdated files... none found
pickling environment... done
checking consistency... /private/tmp/docs/source/foo.rst: WARNING: document isn't included in any toctree
done
preparing documents... done
writing output... [100%] index

waiting for workers...
Process Process-1:
Traceback (most recent call last):
  File "/Users/sfranklin/Python/Python-2.7.13/lib/python2.7/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "/Users/sfranklin/Python/Python-2.7.13/lib/python2.7/multiprocessing/process.py", line 114, in run
    self._target(*self._args, **self._kwargs)
  File "/Users/sfranklin/.virtualenvs/sphinx/lib/python2.7/site-packages/sphinx/util/parallel.py", line 94, in _process
    pipe.send((failed, collector.logs, ret))
PicklingError: Can't pickle <type 'function'>: attribute lookup __builtin__.function failed
```

#### Error logs / results

No error log was created.

#### Expected results

Should successfully build.

### Reproducible project / your project
[source.zip](https://github.com/sphinx-doc/sphinx/files/1533441/source.zip)

This project only has two files (two files necessary to get the workers apparently) and the offending .rst is

```
.. image:: http://www.sphinx-doc.org/en/stable/_static/sphinxheader.png
   :scale: 50 %
```

This prompts a logging message that warns you can't scale a remote image - and it is this LogRecord that is unserializable and prompts the crash. See below for details.

### Environment info
- OS: Mac
- Python version: 2.7.13
- Sphinx version: 1.6.5

### Additional Notes

I can get a debugger and see that the error happens when data is sent back to the pipe in https://github.com/sphinx-doc/sphinx/blob/master/sphinx/util/parallel.py#L94

Trying a pickle.dumps on the `collector.logs` results in a pickling error due to a file object not being serializable. Inspecting the LogRecords I see we have a `.location` attribute which in this case is a docutil image node which holds a pointer to the file.

```
(Pdb) p type(collector.logs[0].location)
<class 'docutils.nodes.image'>
```

I can fix the crash by patching the existing sanitization function at https://github.com/sphinx-doc/sphinx/blob/master/sphinx/util/logging.py#L77 and setting the r.location to None so I'm fairly confident it is indeed the location that holds non-serializable data.

**Comment by tk0miya** [comment]
Fixed by #4286 
Thank you for reporting!

**Comment by simeonf** [comment]
Awesome @tk0miya! Thank you for prompt bugfixing!
